File: server/src/main/java/com/kernith/easyinvoice/service/CustomerService.java
package com.kernith.easyinvoice.service;

import com.kernith.easyinvoice.config.AuthPrincipal;
import com.kernith.easyinvoice.data.dto.customer.CreateCustomerRequest;
import com.kernith.easyinvoice.data.dto.customer.UpdateCustomerRequest;
import com.kernith.easyinvoice.data.model.Company;
import com.kernith.easyinvoice.data.model.Customer;
import com.kernith.easyinvoice.data.model.CustomerStatus;
import com.kernith.easyinvoice.data.model.Quote;
import com.kernith.easyinvoice.data.model.UserRole;
import com.kernith.easyinvoice.data.repository.CompanyRepository;
import com.kernith.easyinvoice.data.repository.CustomerRepository;
import com.kernith.easyinvoice.data.repository.QuoteRepository;
import java.util.List;
import java.util.Locale;
import java.util.Optional;

import com.kernith.easyinvoice.helper.Utils;
import org.springframework.http.HttpStatus;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;
import org.springframework.web.server.ResponseStatusException;

/**
 * Customer use-cases including CRUD, status changes, and related quotes.
 */
@Service
public class CustomerService {

    private final CustomerRepository customerRepository;
    private final CompanyRepository companyRepository;
    private final QuoteRepository quoteRepository;

    /**
     * Creates the service with repositories.
     *
     * @param customerRepository customer repository
     * @param companyRepository company repository
     * @param quoteRepository quote repository
     */
    public CustomerService(
            CustomerRepository customerRepository,
            CompanyRepository companyRepository,
            QuoteRepository quoteRepository
    ) {
        this.customerRepository = customerRepository;
        this.companyRepository = companyRepository;
        this.quoteRepository = quoteRepository;
    }

    /**
     * Creates a new customer for the current company.
     *
     * <p>Lifecycle: validate role, ensure VAT uniqueness, load company, map fields, save.</p>
     *
     * @param request customer creation payload
     * @param principal authenticated principal
     * @return saved customer
     * @throws ResponseStatusException if validation or authorization fails
     */
    public Customer createCustomer(CreateCustomerRequest request, AuthPrincipal principal) {
        Utils.requireRoles(principal, List.of(UserRole.COMPANY_MANAGER));

        Long companyId = Utils.getRequiredCompanyId(principal);
        String vatNumber = normalizeVat(request.vatNumber());

        customerRepository.findByCompanyIdAndVatNumberAndStatus(companyId, vatNumber, CustomerStatus.ACTIVE)
                .ifPresent(existing -> {
                    throw new ResponseStatusException(HttpStatus.CONFLICT, "Vat number already used for this company");
                });

        Optional<Company> company = companyRepository.findById(companyId);
        if (company.isEmpty()) {
            throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "Company specified is not available");
        }

        Customer customer = new Customer(company.get());
        customer.setDisplayName(Utils.normalizeRequired(request.displayName()));
        customer.setLegalName(Utils.trimToNull(request.legalName()));
        customer.setStatus(request.status() == null ? CustomerStatus.ACTIVE : request.status());
        customer.setEmail(normalizeEmail(request.email()));
        customer.setPhone(Utils.trimToNull(request.phone()));
        customer.setVatNumber(vatNumber);
        customer.setPec(normalizeEmail(request.pec()));
        customer.setAddress(Utils.trimToNull(request.address()));
        customer.setCity(Utils.trimToNull(request.city()));
        customer.setPostalCode(Utils.trimToNull(request.postalCode()));
        customer.setCountry(normalizeCountry(request.country()));

        return customerRepository.save(customer);
    }

    /**
     * Lists customers with pagination, sorting, and optional search.
     *
     * @param principal authenticated principal
     * @param page page index (0-based)
     * @param size page size
     * @param sort sort spec (field,dir)
     * @param q optional search query
     * @return page of customers
     * @throws ResponseStatusException if authorization fails
     */
    public Page<Customer> listCustomers(
            AuthPrincipal principal,
            int page,
            int size,
            String sort,
            String q
    ) {
        Utils.requireRoles(principal, List.of(UserRole.COMPANY_MANAGER, UserRole.BACK_OFFICE));

        Long companyId = Utils.getRequiredCompanyId(principal);
        PageRequest pageRequest = toPageRequest(page, size, sort);
        return (q == null || q.isBlank())
                ? customerRepository.findByCompanyIdAndStatus(companyId, CustomerStatus.ACTIVE, pageRequest)
                : customerRepository.searchByCompanyIdAndStatus(companyId, CustomerStatus.ACTIVE, q.trim(), pageRequest);
    }

    /**
     * Retrieves a customer by id for the current company.
     *
     * @param customerId customer identifier
     * @param principal authenticated principal
     * @return optional customer
     * @throws ResponseStatusException if authorization fails
     */
    public Optional<Customer> getCustomer(Long customerId, AuthPrincipal principal) {
        Utils.requireRoles(principal, List.of(UserRole.COMPANY_MANAGER, UserRole.BACK_OFFICE));
        Long companyId = Utils.getRequiredCompanyId(principal);
        return customerRepository.findByIdAndCompanyIdAndStatus(customerId, companyId, CustomerStatus.ACTIVE);
    }

    /**
     * Updates editable fields of a customer.
     *
     * <p>Lifecycle: validate role and customer existence, enforce VAT uniqueness, save.</p>
     *
     * @param customerId customer identifier
     * @param request update payload
     * @param principal authenticated principal
     * @return updated customer
     * @throws ResponseStatusException if validation or authorization fails
     */
    public Customer updateCustomer(Long customerId, UpdateCustomerRequest request, AuthPrincipal principal) {
        Long companyId = Utils.getRequiredCompanyId(principal);
        Optional<Customer> optionalCustomer = getCustomerById(principal, customerId, companyId);
        if (optionalCustomer.isEmpty()) {
            throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "CustomerID " +  customerId + " does not exist");
        }
        Customer customer = optionalCustomer.get();
        if (request.displayName() != null) {
            customer.setDisplayName(Utils.normalizeRequired(request.displayName()));
        }
        if (request.legalName() != null) {
            customer.setLegalName(Utils.trimToNull(request.legalName()));
        }
        if (request.email() != null) {
            customer.setEmail(normalizeEmail(request.email()));
        }
        if (request.phone() != null) {
            customer.setPhone(Utils.trimToNull(request.phone()));
        }
        if (request.vatNumber() != null) {
            String vatNumber = normalizeVat(request.vatNumber());
            if (!vatNumber.equals(customer.getVatNumber())) {
                customerRepository.findByCompanyIdAndVatNumberAndStatus(companyId, vatNumber, CustomerStatus.ACTIVE)
                        .ifPresent(existing -> {
                            throw new ResponseStatusException(HttpStatus.CONFLICT, "Vat number already used for this company");
                        });
                customer.setVatNumber(vatNumber);
            }
        }
        if (request.pec() != null) {
            customer.setPec(normalizeEmail(request.pec()));
        }
        if (request.address() != null) {
            customer.setAddress(Utils.trimToNull(request.address()));
        }
        if (request.city() != null) {
            customer.setCity(Utils.trimToNull(request.city()));
        }
        if (request.postalCode() != null) {
            customer.setPostalCode(Utils.trimToNull(request.postalCode()));
        }
        if (request.country() != null) {
            customer.setCountry(normalizeCountry(request.country()));
        }

        return customerRepository.save(customer);
    }

    /**
     * Soft-deletes a customer.
     *
     * @param customerId customer identifier
     * @param principal authenticated principal
     * @return optional result indicating success
     * @throws ResponseStatusException if authorization fails
     */
    public Optional<Boolean> deleteCustomer(Long customerId, AuthPrincipal principal) {
        return setCustomerStatus(customerId, principal, CustomerStatus.DELETED);
    }

    /**
     * Archives a customer.
     *
     * @param customerId customer identifier
     * @param principal authenticated principal
     * @return optional result indicating success
     * @throws ResponseStatusException if authorization fails
     */
    public Optional<Boolean> archiveCustomer(Long customerId, AuthPrincipal principal) {
        return setCustomerStatus(customerId, principal, CustomerStatus.ARCHIVED);
    }

    /**
     * Restores an archived customer to active.
     *
     * @param customerId customer identifier
     * @param principal authenticated principal
     * @return optional result indicating success
     * @throws ResponseStatusException if authorization fails
     */
    public Optional<Boolean> restoreCustomer(Long customerId, AuthPrincipal principal) {
        return setCustomerStatus(customerId, principal, CustomerStatus.ACTIVE);
    }

    /**
     * Lists quotes related to a customer.
     *
     * @param customerId customer identifier
     * @param principal authenticated principal
     * @return list of quotes
     * @throws ResponseStatusException if authorization fails or customer is missing
     */
    public List<Quote> listCustomerQuotes(Long customerId, AuthPrincipal principal) {
        Utils.requireRoles(principal, List.of(UserRole.COMPANY_MANAGER, UserRole.BACK_OFFICE));
        Long companyId = Utils.getRequiredCompanyId(principal);

        Optional<Customer> optionalCustomer = customerRepository.findByIdAndCompanyIdAndStatus(
                customerId,
                companyId,
                CustomerStatus.ACTIVE
        );
        if (optionalCustomer.isEmpty()) {
            throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "Parameters not valid");
        }

        return quoteRepository.findByCompanyIdAndCustomerIdOrderByIssueDateDesc(companyId, customerId);
    }

    private Optional<Customer> getCustomerById(AuthPrincipal principal, Long customerId, Long companyId) {
        Utils.requireRoles(principal, List.of(UserRole.COMPANY_MANAGER));
        return customerRepository.findByIdAndCompanyIdAndStatus(customerId, companyId,CustomerStatus.ACTIVE);
    }

    private Optional<Boolean> setCustomerStatus(Long customerId, AuthPrincipal principal, CustomerStatus status) {
        Long companyId = Utils.getRequiredCompanyId(principal);
        Optional<Customer> optionalCustomer = getCustomerById(principal, customerId, companyId);
        if (optionalCustomer.isEmpty()) {
            return Optional.empty();
        }
        Customer customer = optionalCustomer.get();
        customer.setStatus(status);
        customerRepository.save(customer);
        return Optional.of(Boolean.TRUE);
    }

    // moved to Utils

    private String normalizeEmail(String email) {
        if (email == null) {
            return null;
        }
        String trimmed = email.trim();
        return trimmed.isEmpty() ? null : trimmed.toLowerCase(Locale.ROOT);
    }

    private String normalizeVat(String vatNumber) {
        return vatNumber == null ? null : vatNumber.trim();
    }

    private String normalizeCountry(String country) {
        if (country == null || country.isBlank()) {
            return "IT";
        }
        return country.trim().toUpperCase(Locale.ROOT);
    }

    private PageRequest toPageRequest(int page, int size, String sort) {
        int safePage = Math.max(page, 0);
        int safeSize = size <= 0 ? 20 : Math.min(size, 100);
        Sort sortSpec = parseSort(sort);
        return PageRequest.of(safePage, safeSize, sortSpec);
    }

    private Sort parseSort(String sort) {
        String raw = (sort == null || sort.isBlank()) ? "displayName,asc" : sort.trim();
        String[] parts = raw.split(",", -1);
        String property = parts[0].trim();
        String direction = parts.length > 1 ? parts[1].trim().toLowerCase(Locale.ROOT) : "asc";

        String mappedProperty = switch (property) {
            case "legalName" -> "legalName";
            case "email" -> "email";
            case "vatNumber" -> "vatNumber";
            case "country" -> "country";
            case "status" -> "status";
            case "createdAt" -> "createdAt";
            case "updatedAt" -> "updatedAt";
            default -> "displayName";
        };

        Sort.Direction sortDirection = "desc".equals(direction) ? Sort.Direction.DESC : Sort.Direction.ASC;
        return Sort.by(sortDirection, mappedProperty).and(Sort.by(Sort.Direction.ASC, "id"));
    }
}

File: client/src/core/services/customer-service.ts
import { BaseService } from "./base-service";
import { Page, PageQuery } from "@/src/core/models/common";

/** Payload for creating a customer. */
export interface CreateCustomerRequest {
  /** Customer display name. */
  displayName: string;
  /** Customer legal name (optional). */
  legalName?: string;
  /** VAT number. */
  vatNumber: string;
  /** Customer email (optional). */
  email?: string;
  /** Phone (optional). */
  phone?: string;
  /** PEC email (optional). */
  pec?: string;
  /** Address (optional). */
  address?: string;
  /** City (optional). */
  city?: string;
  /** Postal code (optional). */
  postalCode?: string;
  /** Country code (optional). */
  country?: string;
  /** Status (optional). */
  status?: "ACTIVE" | "ARCHIVED" | "DELETED";
}

/** Payload for updating a customer. */
export interface UpdateCustomerRequest {
  /** Updated display name. */
  displayName?: string;
  /** Updated legal name. */
  legalName?: string;
  /** Updated VAT number. */
  vatNumber?: string;
  /** Updated email. */
  email?: string;
  /** Updated phone. */
  phone?: string;
  /** Updated PEC. */
  pec?: string;
  /** Updated address. */
  address?: string;
  /** Updated city. */
  city?: string;
  /** Updated postal code. */
  postalCode?: string;
  /** Updated country code. */
  country?: string;
}

/** Summary response for a customer. */
export interface CustomerSummaryResponse {
  /** Customer identifier. */
  id: number;
  /** Customer display name. */
  displayName: string;
  /** Customer legal name (optional). */
  legalName?: string | null;
  /** VAT number (optional). */
  vatNumber?: string | null;
  /** PEC email (optional). */
  pec?: string | null;
  /** Country code (optional). */
  country?: string | null;
}

/** Detailed response for a customer. */
export interface CustomerDetailResponse extends CustomerSummaryResponse {
  /** Status enum string. */
  status?: "ACTIVE" | "ARCHIVED" | "DELETED" | string;
  /** Customer email (optional). */
  email?: string | null;
  /** Phone (optional). */
  phone?: string | null;
  /** Address (optional). */
  address?: string | null;
  /** City (optional). */
  city?: string | null;
  /** Postal code (optional). */
  postalCode?: string | null;
  /** Created date-time. */
  createdAt?: string | null;
  /** Updated date-time. */
  updatedAt?: string | null;
}

/** Summary response for a quote. */
export interface QuoteSummaryResponse {
  /** Quote identifier. */
  id: number;
  /** Quote year. */
  quoteYear?: number | null;
  /** Quote number. */
  quoteNumber?: number | null;
  /** Status enum string. */
  status?: string | null;
  /** Title (optional). */
  title?: string | null;
  /** Issue date (optional). */
  issueDate?: string | null;
  /** Valid until (optional). */
  validUntil?: string | null;
  /** Currency (optional). */
  currency?: string | null;
  /** Total amount (optional). */
  totalAmount?: number | null;
  /** Customer id (optional). */
  customerId?: number | null;
  /** Customer display name (optional). */
  customerDisplayName?: string | null;
}

/** Customer-related API service. */
export class CustomerService extends BaseService {
  /** Base path for customer endpoints. */
  protected basePath(): string {
    return "/manager/customers";
  }

  /**
   * Creates a new customer.
   * @param request creation payload
   */
  createCustomer(request: CreateCustomerRequest): Promise<CustomerSummaryResponse> {
    return this.post<CustomerSummaryResponse, CreateCustomerRequest>(this.endpoint(), { body: request });
  }

  /**
   * Returns a customer by id.
   * @param customerId customer identifier
   */
  getCustomer(customerId: number | string): Promise<CustomerDetailResponse> {
    return this.get<CustomerDetailResponse>(this.endpoint(`/${customerId}`));
  }

  /**
   * Updates a customer by id.
   * @param customerId customer identifier
   * @param request update payload
   */
  updateCustomer(customerId: number | string, request: UpdateCustomerRequest): Promise<CustomerDetailResponse> {
    return this.patch<CustomerDetailResponse, UpdateCustomerRequest>(this.endpoint(`/${customerId}`), {
      body: request,
    });
  }

  /**
   * Deletes (soft-delete) a customer.
   * @param customerId customer identifier
   */
  deleteCustomer(customerId: number | string): Promise<void> {
    return this.delete<void>(this.endpoint(`/${customerId}`));
  }

  /**
   * Lists customers with detail payloads (default).
   * @param params pagination and query params
   */
  listCustomers(params?: PageQuery): Promise<Page<CustomerDetailResponse>>;
  /**
   * Lists customers with summary payloads.
   * @param params pagination and query params with type="summary"
   */
  listCustomers(params: PageQuery & { type: "summary" }): Promise<Page<CustomerSummaryResponse>>;
  /**
   * Lists customers with either summary or detail payloads based on `type`.
   * @param params pagination and query params
   */
  listCustomers(params: PageQuery & { type?: "summary" } = {}): Promise<Page<CustomerDetailResponse> | Page<CustomerSummaryResponse>> {
    const query = { ...params } as Record<string, string | number | boolean | null | undefined>;
    if (params.type === "summary") {
      return this.get<Page<CustomerSummaryResponse>>(this.endpoint(), { query: { ...query, type: "summary" } });
    }
    return this.get<Page<CustomerDetailResponse>>(this.endpoint(), { query });
  }

  /**
   * Archives a customer.
   * @param customerId customer identifier
   */
  archiveCustomer(customerId: number | string): Promise<void> {
    return this.post<void, unknown>(this.endpoint(`/${customerId}/archive`));
  }

  /**
   * Restores an archived customer.
   * @param customerId customer identifier
   */
  restoreCustomer(customerId: number | string): Promise<void> {
    return this.post<void, unknown>(this.endpoint(`/${customerId}/restore`));
  }

  /**
   * Lists quotes for a given customer.
   * @param customerId customer identifier
   */
  listCustomerQuotes(customerId: number | string): Promise<QuoteSummaryResponse[]> {
    return this.get<QuoteSummaryResponse[]>(`/customer/${customerId}/quotes`);
  }
}
